crates/release_plz_core/src/published_packages/git_tags.rs

```diff
@@ -0,0 +1,459 @@
+    ) -> anyhow::Result<Option<impl Summary + 'a>> {
+        // Find the package release tag corresponding to the greatest (i.e. latest) version
+        filter_release_tags(
+            self.tags.iter().map(AsRef::as_ref),
+            package_name,
+            self.project,
+        )
+        .filter_map(|(tag, version)| {
+            self.check_release_tag_validity(tag, package_name, &version)
+                .inner_map(|relative_manifest_path| (tag, version, relative_manifest_path))
+                .transpose()
+        })
+        .process_results(|tags| {
+            tags.max_by(|(_, version1, _), (_, version2, _)| version1.cmp(version2))
+        })
+        .inner_map(|(tag, version, relative_manifest_path)| ReleaseTag {
+            package_name,
+            repo: self.repo,
+            tag,
+            version,
+            relative_manifest_path,
+        })
+    }
+}
+
+/// Filters the release tags for the given package from all the `tags` in a repository.
+///
+/// Each item in the returned iterator is a tuple containing the name of the release tag and the
+/// package version it corresponds to.
+fn filter_release_tags<'t>(
+    tags: impl Iterator<Item = &'t str> + 't,
+    package: &'t str,
+    project: &'t Project,
+) -> impl Iterator<Item = (&'t str, Version)> + 't {
+    // TODO: Consider using git tag template in the release-plz config at each tag, rather than
+    // using the current template
+
+    let version_var = crate::tera::tera_var(crate::tera::VERSION_VAR);
+
+    // By substituting the version variable expression for the version variable,
+    // we only render the package name (if needed) in the template
+    let partial_template = project.git_tag(package, &version_var);
+
+    // Escape the partially rendered template so that it can be used as a regex literal
+    let fully_escaped_partial_template = regex::escape(&partial_template);
+
+    // Replace the escaped version variable expression substring with the original un-escaped
+    // expression string so we can use it as a template
+    let escaped_partial_template =
+        fully_escaped_partial_template.replace(&regex::escape(&version_var), &version_var);
+
+    // Render template with version = "(.+)" to generate a regex which
+    // captures the version string in a group
+    let context = crate::tera::tera_context(package, r"(.+)");
+    let release_tag_regex =
+        crate::tera::render_template(&escaped_partial_template, &context, "release_tag_regex");
+    // Add anchors to ensure regex matches whole string
+    let release_tag_regex =
+        Regex::new(&format!("^{release_tag_regex}$")).expect("invalid rendered version tag regex");
+
+    tags.filter_map(move |tag| {
+        // Check if the tag name matches the regex
+        let (_, [version_str]) = release_tag_regex.captures(tag)?.extract();
+        // Check if the captured version string can be parsed as a package version
+        Some((tag, Version::parse(version_str).ok()?))
```

Comment:
I wonder if we could move some filters like this one when we retrieve the tags for the first time (e.g. in the `get_all_tags` function).

So that we run this filter only once instead of running it for each package. We could gain a bit of performance in this way.

Reply:
This may not be possible with the code as-is - because the git tag template may contain the package name, every package could have a unique `release_tag_regex`. The goal of this filtering is to find the release git tags that correspond to a specific package. 

Perhaps a faster way could be to make the regex capture the package name in addition to the version string - this will let us precompute the list of all release tags for any package in the repository. We will still have to do some per package filtering, but it would simply be a name equality test, which should be faster than compiling a new regex every time.

Reply:
Don't bother if it's too complex of it makes the code way more complex. I thought this was a quick win, but if it's not you can leave the code as it is 👍